### PR TITLE
Adding Signbit & Fmod in ExprEval

### PR DESCRIPTION
**Pull request by jjsjann123**
Signbit & Fmod is used to produce scalar value to reshape, which requires ExpressionEvaluator to run it with scalar value.

Note: Also added the rule in NaiveValueMachine used by PrecomputedValues. But currently those are not being tests.

### CODE REVIEW

File: csrc/ir/nodes.cpp [hunk 1]
```diff
@@ -464,6 +464,17 @@ std::vector<PolymorphicValue> UnaryOp::evaluate(
     case UnaryOpType::Sin:
       return {in.as<at::Tensor>().sin()};
       break;
+    case UnaryOpType::Signbit:
```
Comment (jjsjann123) [inline csrc/ir/nodes.cpp:467]: Question for @Priya2698 .

Are we supposed to not have the full coverage on operations from expr eval ?

File: tests/python/test_python_frontend.py [hunk 2]
```diff
@@ -4344,5 +4344,38 @@ def fusion_func(fd: FusionDefinition):
...
+            S10 = fd.define_scalar(0, dtype=DataType.Int)
+            S11 = fd.ops.ne(S9, S10)
+            S12 = fd.ops.bitwise_and(S6, S11)
+            S13 = fd.define_scalar(192, dtype=DataType.Int)
+            S14 = fd.ops.reciprocal(S3)
+            S15 = fd.ops.mul(S13, S14)
+            S16 = fd.ops.cast(S12, dtype=DataType.Int)
+            S17 = fd.ops.sub(S15, S16)
+            V18 = fd.define_vector([S0, S17], dtype=DataType.Int)
+            T19 = fd.ops.reshape(T1, new_shape=V18)
```
Comment (jjsjann123) [inline tests/python/test_python_frontend.py:4373]: `reshape` concretization here calls evaluate on the chain of operations that produces V18.

signbit/fmod are the two that is missing from what we support, hence I needed to add those. My naive question is, what's the criteria for adding something to what expr eval?

File: csrc/ir/nodes.cpp (hunk 1 above)
Comment (Priya2698) [reply csrc/ir/nodes.cpp:467]: We do intend on having full coverage.
When resolving Issue #670, we targeted developing for all operators seen in test cases. As we add more test cases, and missing operators are seen, we should implement evaluate methods for them.

Feel free to open and assign these issues to me, to address missing operators.

File: tests/python/test_python_frontend.py (hunk 2 above)
Comment (jacobhinkle) [reply tests/python/test_python_frontend.py:4373]: I think we'd like to have evaluate implemented for everything. Note that there is also `NaiveValueMachine::runBinaryOp` and friends in `evaluator_common.cpp`.

Comment (jjsjann123) [reply tests/python/test_python_frontend.py:4373]: Looks like I can just add an entry there. But I'm wondering where can I add a test for the entry in `NaiveValueMachine::runBinaryOp`?

Comment (jacobhinkle) [reply tests/python/test_python_frontend.py:4373]: Good question. If a `PrecomputedValues` is bound to an `ExpressionEvaluator` then you can run `expr_eval.precomputedValues()->evaluate()` as is done here: https://github.com/NVIDIA/Fuser/blob/65799ed56f66b671dfbd51336fa302400e2bdf45/csrc/executor.cpp#L1210. That will run `PrecomputedValues::run()`. It looks like our tests don't actually exercise this very much; the `TernaryOp` test tests that a value can be searched in `PrecomputedValues` but I don't believe that tests the `PrecomputedValues::run()` method: https://github.com/NVIDIA/Fuser/blob/65799ed56f66b671dfbd51336fa302400e2bdf45/tests/cpp/test_evaluator.cpp#L559-L560. That could be added in another PR I think.

Comment (jjsjann123) [reply tests/python/test_python_frontend.py:4373]: I'm going to leave `adding tests for PrecomputedValues` in another PR.

But looks like it makes sense to keep the implementation consistent between the two. So I'll at least add r signbit & fmod and cpp tests for that.

Thanks for pointing this one out. :bow:

Comment (jjsjann123) [reply tests/python/test_python_frontend.py:4381]: I just realized that `PrecomputedValues` are only used for ID extent() here: https://github.com/NVIDIA/Fuser/blob/6c6f3a40e09e6f8bece80b9b79c543945846c71b/csrc/evaluator_common.cpp#L93-L117

I don't see an easy way to get that with `Signbit` or `fmod`. Would you mind if we leave the testing part out? :stuck_out_tongue_closed_eyes:

**Comment by jjsjann123** [comment]
!build

File: csrc/ir/nodes.cpp (hunk 1 above)
Comment (jjsjann123) [reply csrc/ir/nodes.cpp:467]: I'm not trying to push this out as high priority. I think it's reasonably fine to patch it as issues surface.

**Comment by jjsjann123** [comment]
The signal I'm getting is that this is a step towards the right direction. Marked this PR as ready for review.

**Review by Priya2698 — APPROVED** [review]
LGTM. 
Just a minor question about coverage of float type.

File: csrc/polymorphic_value.h [hunk 3]
```diff
@@ -279,6 +279,62 @@ inline bool isSame(const PolymorphicValue& a, const PolymorphicValue& b) {
   return a == b;
 }
 
+inline PolymorphicValue signbit(const PolymorphicValue& a) {
+  if (a.is<int64_t>()) {
+    return PolymorphicValue(std::signbit(a.as<int64_t>()));
+  }
+  if (a.is<double>()) {
+    return PolymorphicValue(std::signbit(a.as<double>()));
+  }
```
[the remark this answers is not in the document]
Comment (jjsjann123) [reply csrc/polymorphic_value.h:288]: IIUC, we don't have `float` for scalars. Same as `int` vs `int64_t`.

**Comment by jjsjann123** [comment]
tests clears despite some numerical issue. I'll merge after CI clear.

**Comment by jjsjann123** [comment]
!build

**Comment by jjsjann123** [comment]
!build

**Comment by jjsjann123** [comment]
This one is low risk and it has previously passed CI earlier. I'm merging it as-is.

**Pull request merged**